### PR TITLE
Add bounds check to MemIo::seek().

### PR DESCRIPTION
**Pull request by kevinbackhouse**
Quick fix for bug reported in #943.

### CODE REVIEW

**Comment by D4N** [comment]
@kevinbackhouse Thanks for the quick fix, the changes look good to me. Would you mind adding a unit test that detects regressions in this behavior?

**Comment by kevinbackhouse** [comment]
Yes, I'll add a test.
I was also thinking that it might be a good idea to add an error flag to `MemIo::Impl`, so that we can implement `MemIo::error()` properly. A lot of the clients of this class call `error()` immediately after `seek()` to check if anything went wrong. Right now, `MemIo::error()` always returns 0.

**Comment by kevinbackhouse** [comment]
I added a unit test. I also changed the bug fix so that it sets the `eof_` flag when you seek out of bounds. I decided against adding an error flag because I am not sure what it would mean.

File: unitTests/test_basicio.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#include <exiv2/basicio.hpp>
+
+#include <cmath>
```
Comment (piponazo) [inline unitTests/test_basicio.cpp]: [minor] These headers seem to not be needed, right?. The only usage from standard functions that I see is `memset` which would be available in `#include <cstring>`

File: unitTests/test_basicio.cpp [hunk 2]
```diff
@@ -19,6 +19,42 @@ TEST(MemIo, seek_out_of_bounds)
...
     // in bounds.
     byte tmp[16];
-    io.seek(0x10000000, BasicIo::beg);
+    ASSERT_EQ(io.seek(0x10000000, BasicIo::beg), 1);
+    ASSERT_TRUE(io.eof());
+
+    // The seek was invalid, so the offset didn't change and this read still works.
+    ASSERT_EQ(io.read(tmp, sizeof(tmp)), sizeof(tmp));
+
+    // Reset the file.
```
Comment (piponazo) [inline unitTests/test_basicio.cpp]: [suggestion] It seems that every time you are resetting, your aim is to make a slightly different test/check. I would propose to create different tests for each case, and try to give meaningful names to each of them.

If that's not the case or you prefer to not make more changes here let me know.

**Review by piponazo — DISMISSED** [review]
Thanks a lot for your contributions!

I left two minor suggestions but I am happy with the PR and we could merge it as it is.

**Review by D4N — COMMENTED** [review]
The changes look good to me, thanks @kevinbackhouse!

Concerning the unit test: I'd prefer to move these into separate functions and to use a fixture for the `MemIo` initialization.

**Review by D4N — CHANGES_REQUESTED** [review]
Please mention CVE-2019-13504 in the commit log, in case anyone wants to cherry-pick fix.

**Comment by clanmills** [comment]
It would be good to get this into 0.27.2.3 (RC3).  To do this, it has be in 0.27-maintenance today (2019-07-14), otherwise it'll have to wait for 0.27.3 (scheduled 2019-09-30).

**Review by piponazo — APPROVED** [review]

**Review by D4N — APPROVED** [review]

**Pull request merged**